### ISSUE TITLE
gardenlet leaks memory

### ISSUE THREAD

**Issue by timebertt** [gardener/gardener#4609]
**How to categorize this issue?**

/area robustness
/kind bug

**What happened**:

In our larger landscapes with 100-260 shoots per Seed we can observe a memory usage pattern of the gardenlet similar to this one:

<img width="2051" alt="Screen Shot 2021-09-06 at 15 12 11" src="https://user-images.githubusercontent.com/46341950/132223233-2b4fbaf2-6d1e-4cfc-a461-9b2ae06c3c27.png">

The green graph shows a strongly increasing memory footprint which gets reset after a restart every once in a while. 
The yellow graph shows gardenlet's memory requests, indicating that VPA is quite confused about gardenlet's behavior... 

This appears to be a typical memory leak.

**What you expected to happen**:

The gardenlet not to leak memory.

**How to reproduce it (as minimally and precisely as possible)**:

Our observations indicate, that the severity of the memory leak increases with more shoots per seed.

**Anything else we need to know?**:

We can observe this pattern for as long as we still have monitoring data on the Seeds. Hence, we can't tell since which version this occurs (but at least since `v1.30.0`).

**Environment**:

- Gardener version: `>= v1.30.0`

**Comment by timebertt** [comment]
@BeckerMax and I were able to find a goroutine leak (which implies a memory leak as well), which is triggered on every `CheckTunnelConnection` operation.
As this is executed by the shoot care controller once per health check operation, we were able to reproduce and provoke it by setting `GardenletConfiguration.controllers.shootCare.syncPeriod=10s` in our local setup.
We were able to verify the leakage with the help of https://github.com/gardener/gardener/pull/4568.

The problematic code is in `CheckForwardPodPort`: https://github.com/gardener/gardener/blob/b04095a46b9ec015657710a094482910bc57e420/pkg/client/kubernetes/pods.go#L115-L128

Because `errChan` is unbuffered, sending to `errChan` in L117 blocks, if there is no receiver is reading from it. As this is the case on almost every call to the function (port forward either gets ready or a timeout occurs so the sole receiver in L122 doesn't read from it anymore), this leaks a goroutine and channel every time (also see [this](https://alenkacz.medium.com/an-example-of-a-goroutine-leak-and-how-to-debug-one-a0697cf677a3) and [this](https://yuriktech.com/2020/11/07/Golang-Memory-Leaks/) blog post).

Currently, we can't tell if this is the only memory leak in gardenlet or whether there are more.
Hence, we will fix it for the time being and observe the behavior on our landscapes once the fix is rolled out.

/assign @BeckerMax @timebertt

**Comment by timebertt** [comment]
This test demonstrates the leak described above: https://github.com/timebertt/gardener/blob/a51b9eabc05b5659fb6dbd6ac078e8ef72fe0f14/pkg/client/kubernetes/leak/leak_test.go

```
$ go test ./pkg/client/kubernetes/leak -memprofile /tmp/mem-leak
ok  	github.com/gardener/gardener/pkg/client/kubernetes/leak	1.411s

$ go test ./pkg/client/kubernetes/leak -memprofile /tmp/mem-fixed
ok  	github.com/gardener/gardener/pkg/client/kubernetes/leak	1.185s

$ go tool pprof -inuse_space -top -diff_base /tmp/mem-fixed /tmp/mem-leak
Type: inuse_space
Time: Sep 6, 2021 at 3:43pm (CEST)
Showing nodes accounting for 57.61MB, 100% of 57.61MB total
      flat  flat%   sum%        cum   cum%
   30.01MB 52.09% 52.09%    30.01MB 52.09%  runtime.malg # <- goroutine allocation (including stacktrace, etc.)
      12MB 20.83% 72.93%       12MB 20.83%  github.com/gardener/gardener/pkg/client/kubernetes/leak.leakGoroutine # <- allocation of `stopChan` and `errChan`
   10.50MB 18.23% 91.16%    14.50MB 25.17%  github.com/gardener/gardener/pkg/client/kubernetes/leak.leakGoroutine.func2 # <- fmt.Errorf("foo")
# ...
```

**Comment by timebertt** [comment]
I quickly checked, whether there is a linter for usage of unbuffered channels, but I couldn't find one.
However, we should at least check, if we have other critical unbuffered channels in our codebase.

**Comment by timebertt** [comment]
I scanned our codebase for the use of unbuffered channels, that could cause a memory leak similar to https://github.com/gardener/gardener/issues/4609#issuecomment-913658722.

We have actually a lot unbuffered channels. However, most of them are not problematic:

- some are never written to, e.g. `stopCh` like [this one](https://github.com/gardener/gardener/blob/975035977a4f776dfe0353c5752d2cfe85c768da/pkg/healthz/periodic.go#L61)
- we can be sure, that receiver is always ready to read, e.g. `workerCh` in controllers ([example](https://github.com/gardener/gardener/blob/841bed4c17895036dac7f5186569e6adf2d377d5/pkg/controllermanager/controller/bastion/bastion.go#L85)) or some result channels from parallel computations/checks ([example](https://github.com/gardener/gardener/blob/712a875b5965c9c87a8ce9c304c13733fe4b11ce/extensions/pkg/controller/healthcheck/healtcheck_actuator.go#L108)) – the receiver is the `for` loop with the `range` operator and the loop never exits / never stops reading until the channel is closed
- `flow.execution.done` (created [here](https://github.com/gardener/gardener/blob/37f306c26c64b810fe81491f8e7f229d76cb8d28/pkg/utils/flow/flow.go#L188))
  - [here](https://github.com/gardener/gardener/blob/37f306c26c64b810fe81491f8e7f229d76cb8d28/pkg/utils/flow/flow.go#L234) we can leak goroutines, if there are duplicated task IDs in the flow, because we don't read from `done` until it is closed but rather rely on the stats [here](https://github.com/gardener/gardener/blob/37f306c26c64b810fe81491f8e7f229d76cb8d28/pkg/utils/flow/flow.go#L295)
  - not problematic, because duplicated task IDs are forbidden: [`Graph.Add` panics](https://github.com/gardener/gardener/blob/1388b5bab6ac235676ad5393634280ba79e64f76/pkg/utils/flow/graph.go#L70)

I found one unbuffered channels that can be problematic:
- `flow.ParallelExitOnError` leaks a goroutine, if it [exits early](https://github.com/gardener/gardener/blob/a1b0891be6061c9511b7d6038e381c4a1a6c0f04/pkg/utils/flow/taskfn.go#L174-L178) (receiver will stop reading before channel is closed)
  - then, [this send operation](https://github.com/gardener/gardener/blob/a1b0891be6061c9511b7d6038e381c4a1a6c0f04/pkg/utils/flow/taskfn.go#L165) will block and leak a goroutine

**Comment by timebertt** [comment]
/unassign @BeckerMax @timebertt 
for now, as all currently planned steps have been completed.
We will have to wait for the fixes to be released and observe their behavior on our internal landscapes.

**Comment by rfranzke** [comment]
@timebertt Can you provide an update here whether the fixes have helped?

**Comment by timebertt** [comment]
Definitely :)

Well, the fixes have helped in a sense, that we don't leak as many goroutines anymore. However, we still see a memory leakage in gardenlet nevertheless.

@ialidzhikov pointed out, that there is a memory leak in the client-go workqueue library, which is fixed in https://github.com/kubernetes/kubernetes/pull/104991. It is included in `v1.23.0` and will be part of the next patch release for `v1.22`. 
In the next step, we can upgrade our dependencies and see if the memory leak still happens.

**Comment by timebertt** [comment]
With gardener v1.44.3 (which includes https://github.com/kubernetes/kubernetes/pull/104991), the memory usage pattern now looks similar to this:

<img width="1653" alt="Screen Shot 2022-04-19 at 09 39 58" src="https://user-images.githubusercontent.com/46341950/163951261-28ab4291-abfc-432b-baec-2a2c97ac444a.png">

The memory usage looks much more stable and doesn't indicate a memory leak so clearly as we saw in the sawtooth pattern in the original issue description.
Hence, I think we can close this issue.
/close
